### PR TITLE
Pydocgen was failing the release build due to dependency errors.

### LINKED ISSUE #8645: [cli] Publish action broken
Releasing v3.56.0 had this error
https://github.com/pulumi/docs/actions/runs/4319115547/jobs/7550261884

```
 pipenv run python -m pydocgen ../../static-prebuilt/docs/reference/pkg/python ../../content/docs/reference/pkg pulumi
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/lib/python3.10/runpy.py", line 146, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/usr/lib/python3.10/runpy.py", line 110, in _get_module_details
    __import__(pkg_name)
  File "/home/runner/work/docs/docs/docs/tools/pydocgen/pydocgen/__init__.py", line 33, in <module>
    from jinja2 import Environment, PackageLoader, select_autoescape
  File "/home/runner/.local/share/virtualenvs/pydocgen-S9NxBeBx/lib/python3.10/site-packages/jinja2/__init__.py", line 12, in <module>
    from .environment import Environment
  File "/home/runner/.local/share/virtualenvs/pydocgen-S9NxBeBx/lib/python3.10/site-packages/jinja2/environment.py", line 25, in <module>
    from .defaults import BLOCK_END_STRING
  File "/home/runner/.local/share/virtualenvs/pydocgen-S9NxBeBx/lib/python3.10/site-packages/jinja2/defaults.py", line 3, in <module>
    from .filters import FILTERS as DEFAULT_FILTERS  # noqa: F401
  File "/home/runner/.local/share/virtualenvs/pydocgen-S9NxBeBx/lib/python3.10/site-packages/jinja2/filters.py", line 13, in <module>
    from markupsafe import soft_unicode
ImportError: cannot import name 'soft_unicode' from 'markupsafe' (/home/runner/.local/share/virtualenvs/pydocgen-S9NxBeBx/lib/python3.10/site-packages/markupsafe/__init__.py)
```

Similar to https://github.com/pulumi/docs/issues/8196

### PR DESCRIPTION
**Pull request by dixler**
Following Workflow failed:
https://github.com/pulumi/docs/actions/runs/4319115547/jobs/7550261884

Pinned m2r < 2.0.0 according to suggestion in https://github.com/Azure/azure-sdk-for-python/issues/22019#issue-1072490557
Bumped Jinja2 to 3.1.2 in Pipenv according to discussion in https://github.com/SRserves85/avro-to-python/issues/20

Fixes #8645

Do I need to do anything else in tandem with resolving the aforementioned error in CI?

### CODE REVIEW

File: content/docs/reference/pkg/python/pulumi/_index.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ notitle: true
 {{< resource-docs-alert "pulumi" >}}
 
 <section id="pulumi-python-sdk">
-<h1>Pulumi Python SDK<a class="headerlink" href="#pulumi-python-sdk" title="Permalink to this headline"></a></h1>
+<h1>Pulumi Python SDK<a class="headerlink" href="#pulumi-python-sdk" title="Permalink to this heading"></a></h1>
```
Comment (dixler) [inline content/docs/reference/pkg/python/pulumi/_index.md:11]: Strange

**Review by justinvp — APPROVED** [review]
LGTM

**Comment by justinvp** [comment]
I'm going to merge this to unblock re-running the job from the release.

**Comment by abhinav** [comment]
@justinvp This shouldn't have been merged yet.

@dixler was working on fixing this artifact
https://www.pulumi.com/docs/reference/pkg/python/pulumi/#pulumi-python-sdk

image

**Comment by abhinav** [comment]
It breaks the header links:
image

Gives back: https://www.pulumi.com/docs/reference/pkg/python/pulumi/#the-pulumi-python-resource-model%EF%83%81

**Pull request merged**